File: javascript/secret-handshake/secret-handshake.js
const keywords = ['wink', 'double blink', 'close your eyes', 'jump'];

export const secretHandshake = (num) => {
    if (typeof num != 'number') {
        throw 'Handshake must be a number';
    }

    let n = num;
    let matchingKeywords = keywords.filter((word, idx) => {
        return n & (1 << idx);
    });

    if (n & (1 << 4)) {
        matchingKeywords = matchingKeywords.reverse();
    }

    return matchingKeywords;
};

File: clojure/secret-handshake/src/secret_handshake.clj
(ns secret-handshake)

(def events [[1 "wink"]
             [2 "double blink"]
             [4 "close your eyes"]
             [8 "jump"]])

(defn- has-digit?
  [num n]
  (pos? (bit-and num n)))

(defn- reverse-on-16
  [num commands]
  (if (has-digit? num 16)
    (reverse commands)
    commands))

(defn commands [num]
  (->> events
       (filter (fn [[n _]]
                 (has-digit? num n)))
       (map second)
       (reverse-on-16 num)))
